packages/core/core/src/migrations/database/5.0.0-discard-drafts.ts

```diff
@@ -0,0 +1,99 @@
+/* eslint-disable no-continue */
+import type { UID } from '@strapi/types';
+import type { Database, Migration } from '@strapi/database';
+import { async, contentTypes } from '@strapi/utils';
+
+type DocumentVersion = { documentId: string; locale: string };
+type Knex = Parameters<Migration['up']>[0];
+
+/**
+ * Load a batch of versions to discard.
+ *
+ * Versions with only a draft version will be ignored.
+ * Only versions with a published version (which always have a draft version) will be discarded.
+ */
+async function* getBatchToDiscard({
+  db,
+  trx,
+  uid,
+  batchSize = 1000,
+}: {
+  db: Database;
+  trx: Knex;
+  uid: string;
+  batchSize?: number;
+}) {
+  let offset = 0;
+  let hasMore = true;
+
+  while (hasMore) {
+    // Look for the published entries to discard
+    const batch: DocumentVersion[] = await db
+      .queryBuilder(uid)
+      .select(['id', 'documentId', 'locale'])
+      .where({ publishedAt: { $ne: null } })
+      .limit(batchSize)
+      .offset(offset)
+      .orderBy('id')
+      .transacting(trx)
+      .execute();
+
+    if (batch.length < batchSize) {
+      hasMore = false;
+    }
+
+    offset += batchSize;
+    yield batch;
+  }
```

Comment:
Just a random comment, but to avoid having to deal with batch etc.. You could use the built-in streaming capabilities of knex

Reply:
Is there a way to get a batch of N elements at once? 
I saw there was an .stream() method which result I could iterate in a for loop, but it was giving me 1 element at a time, and I had to buffer them.

This method returned an array of N elements at once, which I later use to batch update them